botocore/paginate.py

```diff
@@ -27,6 +27,149 @@
+        elif isinstance(data, list):
+            return self._encode_list(data, path)
+        elif isinstance(data, six.binary_type):
+            return self._encode_bytes(data, path)
+        else:
+            return data, []
+
+    def _encode_list(self, data, path):
+        """Encode any bytes in a list, noting the index of what is encoded."""
+        new_data = []
+        encoded = []
+        for i, value in enumerate(data):
+            new_path = path + [i]
+            new_value, new_encoded = self._encode(value, new_path)
+            new_data.append(new_value)
+            encoded.extend(new_encoded)
+        return new_data, encoded
+
+    def _encode_dict(self, data, path):
+        """Encode any bytes in a dict, noting the index of what is encoded."""
+        new_data = {}
+        encoded = []
+        for key, value in data.items():
+            new_path = path + [key]
+            new_value, new_encoded = self._encode(value, new_path)
+            new_data[key] = new_value
+            encoded.extend(new_encoded)
+        return new_data, encoded
+
+    def _encode_bytes(self, data, path):
+        """Base64 encode a byte string."""
+        return base64.b64encode(data).decode('utf-8'), [path]
+
+
+class TokenDecoder(object):
+    """Decodes token strings back into dictionaries.
+
+    This performs the inverse operation to the TokenEncoder, accepting
+    opaque strings and decoding them into a useable form.
+    """
+
+    def decode(self, token):
+        """Decodes an opaque string to a dictionary.
+
+        :type token: str
+        :param token: A token string given by the botocore pagination
+            interface.
+
+        :rtype: dict
+        :returns: A dictionary containing pagination information,
+            particularly the service pagination token(s) but also other boto
+            metadata.
+        """
+        json_string = base64.b64decode(token.encode('utf-8')).decode('utf-8')
+        decoded_token = json.loads(json_string)
+
+        # Remove the encoding metadata as it is read since it will no longer
+        # be needed.
+        encoded_keys = decoded_token.pop('boto_encoded_keys', None)
+        if encoded_keys is None:
+            return decoded_token
+        else:
+            return self._decode(decoded_token, encoded_keys)
+
+    def _decode(self, token, encoded_keys):
+        """Find each encoded value and decode it."""
+        for key in encoded_keys:
+            encoded = self._path_get(token, key)
+            decoded = base64.b64decode(encoded.encode('utf-8'))
+            self._path_set(token, key, decoded)
+        return token
+
+    def _path_get(self, data, path):
+        """Return the nested data at the given path.
+
+        For instance:
+            data = {'foo': ['bar', 'baz']}
+            path = ['foo', 0]
+            ==> 'bar'
+        """
+        # jmespath isn't used here because it would be difficult to actually
```

Comment:
Yeah... it would be great if reverse jmespath was a thing. This is good for now.